### ISSUE TITLE
getAs and usingCaching do not function together

### ISSUE THREAD

**Issue by seanmarthur** [SharePoint/PnP-JS-Core#576]
#### Category
[ ] Enhancement

[ X ] Bug

[ ] Question

#### Version

Please specify what version of the library you are using: [ 2.0.8 ]

#### Expected / Desired Behavior / Question

I want to cache the result of a call (_api/web/currentuser) but use getAs to extend the base SiteUser class and imeplement my own properties/methods. The first call (where the value is initially cached) will return the object as an instance of my class. Subsequent calls (cache hits) are returned as a plain object, without my instance properties and functions.

#### Observed Behavior
Cached objects are not cast to an instance of the proper type.

#### Steps to Reproduce
1. Given a class like this: 
```
import { SiteUserProps } from 'sp-pnp-js/lib/sharepoint/siteusers';
import { Item } from 'sp-pnp-js/lib/sharepoint';
export default class User extends Item implements SiteUserProps  {
  // implemented SiteUserProps fields excluded
   public myInstanceFn() {
    alert('instance function!')
  }
}
```
2. Make a call like this: 
`web.currentUser.usingCaching().getAs(ODataEntity(User)).then(user => { user.myInstanceFn() });`

3. After the request is cached, make the exact call again, it will fail: 
`web.currentUser.usingCaching().getAs(ODataEntity(User)).then(user => { user.myInstanceFn() //Exception! 'TypeError: user.myInstanceFn is not a function' });`

The issue seems to stem from the fact that only the data is cached, but the parser seems to rely on the fetch response object.

**Comment by patrick-rodgers** [comment]
This is partly a duplicate of the issue in #390. I'll have a look and see if we can do something in the pipeline to address this.

**Comment by patrick-rodgers** [comment]
I've looked at this and understand the issue - but I am not going to fix it for the 3.0.0 release. I need to figure out a generic way to handle it and don't want to risk tearing up the pipeline as I need to do the release. So...after I get the 3.0.0 release out I will take another look at this and see how best to solve it.

And apologies for the delay in responding, lots happening behind the scenes with refactoring work.

**Comment by patrick-rodgers** [comment]
This is fixed in #615 - sorry for the delay but had to think about the best way to address it as well as balance with other priorities. Will be part of upcoming 3.0.2 release. Thanks!
